File: Project_Graph & Tree_Traversing_Problems/Bài_toán_4.py
def matrix_to_list(matrix):
    adj_list = {}
    for i, row in enumerate(matrix):
        adj_list[i] = [j for j, val in enumerate(row) if val]
    return adj_list

def list_to_matrix(adj_list):
    n = len(adj_list)
    matrix = [[0]*n for _ in range(n)]
    for i, neighbors in adj_list.items():
        for j in neighbors:
            matrix[i][j] = 1
    return matrix

matrix = [
    [0,1,0,1],
    [1,0,1,0],
    [0,1,0,1],
    [1,0,1,0]
]

adj_list = matrix_to_list(matrix)
print("Adjacency List:", adj_list)

new_matrix = list_to_matrix(adj_list)
print("Back to Matrix:")
for row in new_matrix:
    print(row)

File: Project_Graph & Tree_Traversing_Problems/Bài_toán_4.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<vector<int>> matrix_to_list(const vector<vector<int>>& matrix) {
    int n = matrix.size();
    vector<vector<int>> adjList(n);
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            if (matrix[i][j])
                adjList[i].push_back(j);
    return adjList;
}

vector<vector<int>> list_to_matrix(const vector<vector<int>>& adjList) {
    int n = adjList.size();
    vector<vector<int>> matrix(n, vector<int>(n, 0));
    for (int i = 0; i < n; ++i)
        for (int j : adjList[i])
            matrix[i][j] = 1;
    return matrix;
}

int main() {
    vector<vector<int>> matrix = {
        {0,1,0,1},
        {1,0,1,0},
        {0,1,0,1},
        {1,0,1,0}
    };

    auto adjList = matrix_to_list(matrix);
    cout << "Adjacency List:\n";
    for (int i = 0; i < adjList.size(); ++i) {
        cout << i << ": ";
        for (int j : adjList[i]) cout << j << " ";
        cout << "\n";
    }

    auto backMatrix = list_to_matrix(adjList);
    cout << "\nBack to Matrix:\n";
    for (auto& row : backMatrix) {
        for (int val : row) cout << val << " ";
        cout << "\n";
    }

    return 0;
}
